File: swarmmanagercontroller/src/main/java/com/swarmmanager/docker/api/common/client/jaxrs/tls/SSLContextFactory.java
package com.swarmmanager.docker.api.common.client.jaxrs.tls;

import com.swarmmanager.docker.api.common.client.DockerWebClientProperties;
import com.swarmmanager.exception.UnsupportedConfiguration;
import org.apache.log4j.Logger;
import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
import org.bouncycastle.cert.X509CertificateHolder;
import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.openssl.PEMKeyPair;
import org.bouncycastle.openssl.PEMParser;
import org.glassfish.jersey.SslConfigurator;

import javax.net.ssl.SSLContext;
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.*;
import java.security.cert.*;
import java.security.cert.Certificate;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.PKCS8EncodedKeySpec;
import java.util.ArrayList;
import java.util.List;

public class SSLContextFactory {

    private static final Logger LOGGER = Logger.getLogger(SSLContextFactory.class.getName());

    private static final String INTERNAL_KEYSTORE_PASS = "docker";

    private SSLContextFactory() {
    }

    public static SSLContext createSSLContext(DockerWebClientProperties properties) {
        try {
            TLSAuth tlsAuth = TLSAuth.getTLSMode(properties);
            Security.addProvider(new BouncyCastleProvider());
            switch (tlsAuth) {
                case TLS_AUTHENTICATE_CLIENT:
                    return createAuthClientSSLContext(properties.getDockerApiTlsCert().get(), properties.getDockerApiTlsKey().get());
                case TLS_AUTHENTICATE_SERVER:
                    return createAuthServerSSLContext(properties.getDockerApiTlsCacert().get());
                case TLS_AUTHENTICATE_SERVER_CLIENT:
                    return createAuthServerClientSSLContext(properties.getDockerApiTlsCert().get(),
                            properties.getDockerApiTlsKey().get(),
                            properties.getDockerApiTlsCacert().get());
            }
            return SSLContext.getDefault();
        } catch (Exception e) {
            LOGGER.error("Exception creating ssl config ", e);
            throw new UnsupportedConfiguration(e.getMessage());
        }
    }

    private static SSLContext createAuthClientSSLContext(String certPath, String keyPath) throws InvalidKeySpecException,
            CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException {
        SslConfigurator sslConfig = createSSLConfigurator();
        sslConfig.keyStore(createKeyStore(certPath, keyPath));
        sslConfig.keyStorePassword(INTERNAL_KEYSTORE_PASS);
        return sslConfig.createSSLContext();
    }

    private static SSLContext createAuthServerSSLContext(String caCert) throws InvalidKeySpecException,
            CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException {
        SslConfigurator sslConfig = createSSLConfigurator();
        sslConfig.trustStore(createTrustStore(Paths.get(caCert)));
        return sslConfig.createSSLContext();
    }

    private static SSLContext createAuthServerClientSSLContext(String certPath, String keyPath, String caCert) throws InvalidKeySpecException,
            CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException {
        SslConfigurator sslConfig = createSSLConfigurator();
        sslConfig.keyStore(createKeyStore(certPath, keyPath));
        sslConfig.keyStorePassword(INTERNAL_KEYSTORE_PASS);
        sslConfig.trustStore(createTrustStore(Paths.get(caCert)));
        return sslConfig.createSSLContext();
    }

    private static SslConfigurator createSSLConfigurator() {
        SslConfigurator sslConfig = SslConfigurator.newInstance(true);
        sslConfig.securityProtocol("TLSv1.2");
        return sslConfig;
    }

    private static KeyStore createKeyStore(String certPath, String keyPath) throws NoSuchAlgorithmException,
            InvalidKeySpecException, IOException, CertificateException, KeyStoreException {
        PrivateKey privateKey = loadPrivateKey(Paths.get(keyPath));
        List<java.security.cert.Certificate> privateCertificates = loadCertificates(Paths.get(certPath));

        KeyStore keyStore = KeyStore.getInstance("JKS");
        keyStore.load(null);

        keyStore.setKeyEntry(INTERNAL_KEYSTORE_PASS,
                privateKey,
                INTERNAL_KEYSTORE_PASS.toCharArray(),
                privateCertificates.toArray(new java.security.cert.Certificate[privateCertificates.size()])
        );

        return keyStore;
    }

    private static PrivateKey loadPrivateKey(Path keyPath) throws IOException, NoSuchAlgorithmException,
            InvalidKeySpecException {
        if (Files.exists(keyPath)) {

            String key = new String(Files.readAllBytes(keyPath));
            try (StringReader reader = new StringReader(key); PEMParser pemParser = new PEMParser(reader)) {
                Object readObject = pemParser.readObject();
                while (readObject != null) {
                    if (readObject instanceof PEMKeyPair) {
                        PEMKeyPair pemKeyPair = (PEMKeyPair) readObject;
                        PrivateKey privateKey = guessKey(pemKeyPair.getPrivateKeyInfo().getEncoded());
                        if (privateKey != null) {
                            return privateKey;
                        }
                    } else if (readObject instanceof PrivateKeyInfo) {
                        PrivateKeyInfo privateKeyInfo = (PrivateKeyInfo) readObject;
                        PrivateKey privateKey = guessKey(privateKeyInfo.getEncoded());
                        if (privateKey != null) {
                            return privateKey;
                        }
                    }

                    readObject = pemParser.readObject();
                }
            }
        } else {
            throw new UnsupportedConfiguration("Key " + keyPath + " doesn't exist");
        }

        return null;
    }

    private static List<java.security.cert.Certificate> loadCertificates(Path certPath) throws IOException,
            CertificateException {
        if (Files.exists(certPath)) {

            String cert = new String(Files.readAllBytes(certPath));
            try (StringReader reader = new StringReader(cert); PEMParser pemParser = new PEMParser(reader)) {
                List<java.security.cert.Certificate> certificates = new ArrayList<>();

                JcaX509CertificateConverter certificateConverter = new JcaX509CertificateConverter()
                        .setProvider(BouncyCastleProvider.PROVIDER_NAME);
                Object certObj = pemParser.readObject();

                if (certObj instanceof X509CertificateHolder) {
                    X509CertificateHolder certificateHolder = (X509CertificateHolder) certObj;
                    certificates.add(certificateConverter.getCertificate(certificateHolder));
                }

                return certificates;
            }
        } else {
            throw new UnsupportedConfiguration("Certificate " + certPath + " doesn't exist");
        }
    }

    private static PrivateKey guessKey(byte[] encodedKey) throws NoSuchAlgorithmException {
        //no way to know, so iterate
        for (String guessFactory : new String[]{"RSA", "ECDSA"}) {
            try {
                KeyFactory factory = KeyFactory.getInstance(guessFactory);

                PKCS8EncodedKeySpec privateKeySpec = new PKCS8EncodedKeySpec(encodedKey);
                return factory.generatePrivate(privateKeySpec);
            } catch (InvalidKeySpecException ignore) {
            }
        }

        return null;
    }

    private static KeyStore createTrustStore(Path caCertPath) throws IOException, CertificateException,
            KeyStoreException, NoSuchAlgorithmException {
        if (Files.exists(caCertPath)) {

            String caCert = new String(Files.readAllBytes(caCertPath));
            try (StringReader reader = new StringReader(caCert); PEMParser pemParser = new PEMParser(reader)) {

                KeyStore trustStore = KeyStore.getInstance("JKS");
                trustStore.load(null);

                int index = 1;
                Object pemCert;

                while ((pemCert = pemParser.readObject()) != null) {
                    Certificate caCertificate = new JcaX509CertificateConverter()
                            .setProvider(BouncyCastleProvider.PROVIDER_NAME)
                            .getCertificate((X509CertificateHolder) pemCert);
                    trustStore.setCertificateEntry("ca-" + index, caCertificate);
                    index++;
                }

                return trustStore;
            }
        } else {
            throw new UnsupportedConfiguration("CA Certificate " + caCertPath + " doesn't exist");
        }
    }
}

File: swarmmanagercontroller-new/src/main/kotlin/com/swarmbit/docker/api/client/jaxrs/SSLContextFactory.kt
package com.swarmbit.docker.api.client.jaxrs

import com.swarmbit.docker.api.config.DockerClientConfig
import com.swarmbit.docker.api.exception.UnsupportedConfiguration
import org.apache.log4j.Logger
import org.bouncycastle.cert.X509CertificateHolder
import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.bouncycastle.openssl.PEMKeyPair
import org.bouncycastle.openssl.PEMParser
import org.glassfish.jersey.SslConfigurator
import java.io.IOException
import java.io.StringReader
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.security.*
import java.security.cert.Certificate
import java.security.cert.CertificateException
import java.security.spec.InvalidKeySpecException
import java.security.spec.PKCS8EncodedKeySpec
import java.util.*
import javax.net.ssl.SSLContext

object SSLContextFactory {
    private val LOGGER = Logger.getLogger(SSLContextFactory::class.java.name)
    private const val INTERNAL_KEYSTORE_PASS = "docker"

    fun createSSLContext(config: DockerClientConfig): SSLContext {
        return try {
            val tlsAuth: TLSAuth = TLSAuth.getTLSMode(config)
            Security.addProvider(BouncyCastleProvider())
            when (tlsAuth) {
                TLSAuth.TLS_AUTHENTICATE_CLIENT -> return createAuthClientSSLContext(config.tlsCert.orEmpty(), config.tlsKey.orEmpty())
                TLSAuth.TLS_AUTHENTICATE_SERVER -> return createAuthServerSSLContext(config.tlsCacert.orEmpty())
                TLSAuth.TLS_AUTHENTICATE_SERVER_CLIENT -> return createAuthServerClientSSLContext(config.tlsCert.orEmpty(),
                        config.tlsKey.orEmpty(),
                        config.tlsCacert.orEmpty())
                else -> SSLContext.getDefault()
            }
        } catch (e: Exception) {
            LOGGER.error("Exception creating ssl config ", e)
            throw UnsupportedConfiguration(e.message.orEmpty())
        }
    }

    @Throws(InvalidKeySpecException::class, CertificateException::class, NoSuchAlgorithmException::class, KeyStoreException::class, IOException::class)
    private fun createAuthClientSSLContext(certPath: String, keyPath: String): SSLContext {
        val sslConfig = createSSLConfigurator()
        sslConfig.keyStore(createKeyStore(certPath, keyPath))
        sslConfig.keyStorePassword(INTERNAL_KEYSTORE_PASS)
        return sslConfig.createSSLContext()
    }

    @Throws(InvalidKeySpecException::class, CertificateException::class, NoSuchAlgorithmException::class, KeyStoreException::class, IOException::class)
    private fun createAuthServerSSLContext(caCert: String): SSLContext {
        val sslConfig = createSSLConfigurator()
        sslConfig.trustStore(createTrustStore(Paths.get(caCert)))
        return sslConfig.createSSLContext()
    }

    @Throws(InvalidKeySpecException::class, CertificateException::class, NoSuchAlgorithmException::class, KeyStoreException::class, IOException::class)
    private fun createAuthServerClientSSLContext(certPath: String, keyPath: String, caCert: String): SSLContext {
        val sslConfig = createSSLConfigurator()
        sslConfig.keyStore(createKeyStore(certPath, keyPath))
        sslConfig.keyStorePassword(INTERNAL_KEYSTORE_PASS)
        sslConfig.trustStore(createTrustStore(Paths.get(caCert)))
        return sslConfig.createSSLContext()
    }

    private fun createSSLConfigurator(): SslConfigurator {
        val sslConfig = SslConfigurator.newInstance(true)
        sslConfig.securityProtocol("TLSv1.2")
        return sslConfig
    }

    @Throws(NoSuchAlgorithmException::class, InvalidKeySpecException::class, IOException::class, CertificateException::class, KeyStoreException::class)
    private fun createKeyStore(certPath: String, keyPath: String): KeyStore {
        val privateKey = loadPrivateKey(Paths.get(keyPath))
        val privateCertificates = loadCertificates(Paths.get(certPath))
        val keyStore = KeyStore.getInstance("JKS")
        keyStore.load(null)
        keyStore.setKeyEntry(INTERNAL_KEYSTORE_PASS,
                privateKey,
                INTERNAL_KEYSTORE_PASS.toCharArray(),
                privateCertificates.toTypedArray()
        )
        return keyStore
    }

    @Throws(IOException::class, NoSuchAlgorithmException::class, InvalidKeySpecException::class)
    private fun loadPrivateKey(keyPath: Path): PrivateKey? {
        if (Files.exists(keyPath)) {
            val key = String(Files.readAllBytes(keyPath))
            StringReader(key).use { reader ->
                PEMParser(reader).use { pemParser ->
                    var readObject: Any? = pemParser.readObject()
                    while (readObject != null) {
                        if (readObject is PEMKeyPair) {
                            val pemKeyPair: PEMKeyPair = readObject
                            val privateKey = guessKey(pemKeyPair.privateKeyInfo.encoded)
                            if (privateKey != null) {
                                return privateKey
                            }
                        } else if (readObject is org.bouncycastle.asn1.pkcs.PrivateKeyInfo) {
                            val privateKeyInfo: org.bouncycastle.asn1.pkcs.PrivateKeyInfo = readObject
                            val privateKey = guessKey(privateKeyInfo.encoded)
                            if (privateKey != null) {
                                return privateKey
                            }
                        }
                        readObject = pemParser.readObject()
                    }
                }
            }
        } else {
            throw UnsupportedConfiguration("Key $keyPath doesn't exist")
        }
        return null
    }

    @Throws(IOException::class, CertificateException::class)
    private fun loadCertificates(certPath: Path): List<Certificate> {
        if (Files.exists(certPath)) {
            val cert = String(Files.readAllBytes(certPath))
            StringReader(cert).use { reader ->
                PEMParser(reader).use { pemParser ->
                    val certificates: MutableList<Certificate> = ArrayList()
                    val certificateConverter: JcaX509CertificateConverter = JcaX509CertificateConverter()
                            .setProvider(BouncyCastleProvider.PROVIDER_NAME)
                    val certObj: Any = pemParser.readObject()
                    if (certObj is X509CertificateHolder) {
                        val certificateHolder: X509CertificateHolder = certObj
                        certificates.add(certificateConverter.getCertificate(certificateHolder))
                    }
                    return certificates
                }
            }
        } else {
            throw UnsupportedConfiguration("Certificate $certPath doesn't exist")
        }
    }

    @Throws(NoSuchAlgorithmException::class)
    private fun guessKey(encodedKey: ByteArray): PrivateKey? {
        //no way to know, so iterate
        for (guessFactory in arrayOf("RSA", "ECDSA")) {
            try {
                val factory = KeyFactory.getInstance(guessFactory)
                val privateKeySpec = PKCS8EncodedKeySpec(encodedKey)
                return factory.generatePrivate(privateKeySpec)
            } catch (ignore: InvalidKeySpecException) {
            }
        }
        return null
    }

    @Throws(IOException::class, CertificateException::class, KeyStoreException::class, NoSuchAlgorithmException::class)
    private fun createTrustStore(caCertPath: Path): KeyStore {
        if (Files.exists(caCertPath)) {
            val caCert = String(Files.readAllBytes(caCertPath))
            StringReader(caCert).use { reader ->
                PEMParser(reader).use { pemParser ->
                    val trustStore = KeyStore.getInstance("JKS")
                    trustStore.load(null)
                    var index = 1
                    var pemCert: Any?
                    while (pemParser.readObject().also { pemCert = it } != null) {
                        val caCertificate: Certificate = JcaX509CertificateConverter()
                                .setProvider(BouncyCastleProvider.PROVIDER_NAME)
                                .getCertificate(pemCert as X509CertificateHolder?)
                        trustStore.setCertificateEntry("ca-$index", caCertificate)
                        index++
                    }
                    return trustStore
                }
            }
        } else {
            throw UnsupportedConfiguration("CA Certificate $caCertPath doesn't exist")
        }
    }
}